### PR TITLE
feat(docker): add user-provided hooks support to Docker API

### LINKED ISSUE #1377: [Feature]: Add hooks for Docker API
### crawl4ai version

0.7.2

### Expected Behavior

Add `hooks` feature to the Docker API, so users can customize the crawler at specific points in the pipeline.

Reference Discord message: https://discord.com/channels/1278297938551902308/1358770371138031647/1402135668771258378

### Current Behavior

-

### Is this reproducible?

Yes

### Inputs Causing the Bug

```bash

```

### Steps to Reproduce

```bash

```

### Code snippets

```python

```

### OS

Windows, macOS, Linux

### Python version

3.8 +

### Browser

_No response_

### Browser version

_No response_

### Error logs & Screenshots (if applicable)

_No response_

### PR DESCRIPTION
**Pull request by ntohidi**
## Summary
Adds comprehensive user-provided hooks support to the Docker API, allowing users to inject custom Python functions as strings that execute at specific points in the crawling pipeline. This enables advanced customization of crawler behavior without modifying the core codebase.

Fixes #1377

## List of files changed and why
- `deploy/docker/hook_manager.py` - New module implementing UserHookManager for safe validation, compilation, and execution of user-provided hook functions with error isolation
- `deploy/docker/api.py` - Updated handle_crawl_request and handle_stream_crawl_request to integrate hooks, maintain hook_manager instance for execution tracking
- `deploy/docker/schemas.py` - Added HookConfig and CrawlRequestWithHooks models to support hooks in API requests
- `deploy/docker/server.py` - Updated /crawl and /crawl/stream endpoints to accept hooks parameter, added /hooks/info endpoint for hook discovery
- `docs/md_v2/core/docker-deployment.md` - Added comprehensive hooks documentation section with security warnings, real-world examples, and best practices
- `docs/examples/docker_hooks_examples.py` - Created example file demonstrating all 8 hooks with practical use cases (authentication, performance optimization, content extraction)
- `tests/docker/test_hooks_client.py` - Added test suite for basic hooks functionality and error handling
- `tests/docker/test_hooks_comprehensive.py` - Added comprehensive test suite covering all hook points, timeout handling, and edge cases

## How Has This Been Tested?
- Built Docker image locally using `docker compose build`
- Ran container on port 11235 and verified all endpoints
- Executed comprehensive test suite (`docs/examples/docker_hooks_examples.py`) with 100% success rate
- Tested all 8 hook points individually and in combination
- Verified error isolation - malformed hooks don't crash the crawler
- Tested timeout protection with slow-running hooks
- Validated execution tracking and statistics reporting
- Confirmed JSON serialization of all hook results
- Tested with real URLs (httpbin.org, Wikipedia, BBC, GitHub)

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added/updated unit tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: docs/md_v2/core/docker-deployment.md [hunk 1]
```diff
@@ -405,6 +405,409 @@ Executes JavaScript snippets on the specified URL and returns the full crawl res
...
+            if (header && value) {
+                data[header.innerText.trim()] = value.innerText.trim();
+            }
+        });
+        
+        return data;
+    }''')
+    
+    if metadata:
+        console.log("Extracted metadata:", metadata)
```
Comment (SohamKukreti) [inline docs/md_v2/core/docker-deployment.md:662]: Should be `print("Extracted metadata:", metadata)`

**Comment by SohamKukreti** [comment]
I've tested the hooks functionality and everything is working correctly:

1. **API Discovery**: `/hooks/info` endpoint - ✅ Working
2. **Basic Hook Execution**: `on_page_context_created` hook - ✅ Working  
   - Response: `'hooks': {'status': {'status': 'success', 'attached_hooks': ['on_page_context_created']}`
3. **Error Handling**: Malformed hooks - ✅ Working (properly handles errors)
   - Response: `'validation_errors': [{'hook_point': 'invalid_hook_point', 'error': 'Unknown hook point. Valid points: on_browser_created, on_page_context_created, before_goto, after_goto, on_user_agent_updated, on_execution_started, before_retrieve_html, before_return_html', 'code_preview': 'invalid code'}]`
4. **Comprehensive Testing**: `docs/examples/docker_hooks_example.py` - ✅ Working
5. **Authentication Scenario**: Hook-based auth flow - ✅ Working
   - Test: Basic auth with `before_goto` hook adding Authorization header
   - Response: `"authenticated": true, "user": "user"` ✅ Auth successful
   - Hook execution: 0.004s, 100% success rate
6. **Performance Impact Test**: BBC.com with/without hooks - ✅ Working
   - With hook: 3.01s processing, 0.90MB memory delta
   - Without hook: 2.94s processing, 1.46MB memory delta
   - Result: Hook reduces memory usage by 38% (resource blocking effective)
7. **Content Extraction**: Wikipedia metadata extraction - ✅ Working  
   - Test: `after_goto` and `before_retrieve_html` hooks on Wikipedia
   - Successfully extracted Python programming language metadata from infobox  
   - Result: Complete infobox data including paradigm, developer, release info, etc.
   - Hook execution: Both hooks completed successfully (0.028s + 0.018s)


### Assessment

The implementation is solid with:
- Proper error isolation (malformed hooks don't crash crawls)
- Comprehensive validation and security measures
- All 8 hook points functioning correctly
- Good documentation and examples

**LGTM!**
